lib/worker/app-worker.js

```diff
@@ -70,35 +66,76 @@ function launchWorker_p(appSpec, endpoint, logFilePath, workerId) {
     return Q.reject(new Error("No app directory specified"));
 
 
-  return exists_p(appSpec.appDir).then(function(exists) {
+  return exists_p(appSpec.appDir).then(function(exists) { // TODO: does this need to be as user?
     if (!exists) {
       var err = new Error("App dir " + appSpec.appDir + " does not exist");
       err.code = 'ENOTFOUND';
       throw err;
     }
-    
-    // Open the log file asynchronously, then create the worker
-    return Q.nfcall(fs.open, logFilePath, 'a', 0660).then(function(logStream) {
-      fs.fchown(logStream, pw.uid, pw.gid, function(err) {
-        if (err)
-          logger.error('Error attempting to change permissions on log file at ' + logFilePath + ': ' + err.message);
-      });
+   
+    if (!appSpec.logAsUser){
+      // Ensure that the log directory exists.
+      try {
+        fs.mkdirSync(logDir, '755');
+        fs.chownSync(logDir, pw.uid, pw.gid);
+      } catch (ex) {
+        try {
+          var stat = fs.statSync(logDir);
+          if (!stat.isDirectory()) {
+            logger.error('Log directory existed, was a file: ' + logDir);
+            logDir = null;
+          }
+        } catch (ex2) {
+          logger.error('Log directory creation failed: ' + ex2.message);
+          logDir = null;
+        }
+      }
 
-      // Create the worker; when it exits (or fails to start), close
-      // the logStream.
-      var worker = new AppWorker(appSpec, endpoint, logStream, workerId, 
-        pw.home);
-      worker.getExit_p()
-      .fin(function() {
-        fs.close(logStream, function(err) {
+      // Manage the log file as root
+      // Open the log file asynchronously, then create the worker
+      return Q.nfcall(fs.open, logFilePath, 'a', 0660).then(function(logStream) {
+        fs.fchown(logStream, pw.uid, pw.gid, function(err) {
           if (err)
-            logger.error("Couldn't close logStream: " + err.message);
+            logger.error('Error attempting to change permissions on log file at ' + logFilePath + ': ' + err.message);
         });
-      })
-      .eat();
 
-      return worker;
-    });
+        // Create the worker; when it exits (or fails to start), close
+        // the logStream.
+        var worker = new AppWorker(appSpec, endpoint, logStream, workerId, 
+          pw.home);
+        worker.getExit_p()
+        .fin(function() {
+          fs.close(logStream, function(err) {
+            if (err)
+              logger.error("Couldn't close logStream: " + err.message);
+          });
+        })
+        .eat();
+
+        return worker;
+      });
+    } else {
+      var prom = Q.defer();
```

Comment:
I prefer a style where some of this ancillary stuff is hoisted into helper functions rather than making a big function bigger.